cloudinit/sources/DataSourceAeza.py

```diff
@@ -0,0 +1,172 @@
+                raise e
+            # Wait before retrying if retries are left
+            time.sleep(sec_between_retries)
+            attempt += 1
+        # If max_wait timeout is enabled check for max wait timeout
+        if max_wait_seconds is not None:
+            if start + max_wait_seconds < time.time():
+                raise RequestException(
+                    f"Failed to retrieve data from {url} after {retries + 1} attempts because of max wait timeout."
+                )
+
+    # If retries are exhausted, raise an exception indicating the failure
+    raise RequestException(f"Failed to retrieve data from {url} after {retries + 1} attempts.")
+
+
+def read_metadata(url, timeout_seconds=2, max_wait_seconds=120, sec_between_retries=2, retries=30):
+    """
+    Read user data from URL or raise error.
+    """
+    response = read_url(
+        url,
+        timeout_seconds=timeout_seconds,
+        max_wait_seconds=max_wait_seconds,
+        sec_between_retries=sec_between_retries,
+        retries=retries,
+    )
+    if not response.ok:
+        raise RuntimeError("unable to read metadata at %s" % url)
+    return util.load_json(response.content.decode())
+
+
+def read_data(url, timeout_seconds=2, max_wait_seconds=120, sec_between_retries=2, retries=30):
+    """
+    Read data from URL.
+    """
+    try:
+        response = read_url(
+            url,
+            timeout_seconds=timeout_seconds,
+            max_wait_seconds=max_wait_seconds,
+            sec_between_retries=sec_between_retries,
+            retries=retries,
+        )
+        if not response.ok:
+            raise RuntimeError("unable to read data at %s" % url)
+        return response.content
+    except HTTPError as e:
+        if e.response.status_code == 404:
+            return None
+        raise e
+
+
+def format_url(url, system_uuid=None):
+    if system_uuid is None:
+        system_uuid = dmi.read_dmi_data("system-uuid")
+    return url.format(id=system_uuid)
+
+
+class DataSourceAeza(sources.DataSource):
+
+    dsname = "Aeza"
+
+    def __init__(self, sys_cfg, distro, paths, ud_proc=None):
+        super().__init__(sys_cfg, distro, paths, ud_proc)
+
+        self.ds_cfg = util.mergemanydict([self.ds_cfg, BUILTIN_DS_CONFIG])
+
+        url_params = self.get_url_params()
+        self.timeout_seconds = url_params.timeout_seconds
+        self.max_wait_seconds = url_params.max_wait_seconds
+        self.retries = url_params.num_retries
+        self.sec_between_retries = url_params.sec_between_retries
+
+        self.metadata_address = self.ds_cfg["metadata_url"]
+        self.userdata_address = self.ds_cfg["userdata_url"]
+        self.vendordata_address = self.ds_cfg["vendordata_url"]
+
+    @staticmethod
+    def ds_detect():
+        return dmi.read_dmi_data("system-manufacturer") == "Aeza"
+
+    def _get_data(self):
+        system_uuid = dmi.read_dmi_data("system-uuid")
+
+        md = read_metadata(
```

Comment:
The way read_metadata is written, 404s in any of these route returns a None value.

I don't think we generally want to allow for absent or optional metadata or user-data endpoints to have a correctly discovered datasource. 

For cloud-init to be certain it has contacted a viable instance metadata service that matches the discovered datasource type and not some bogus service that happened to be stood up at this public IP address, each datasource generally requires  minimally meta-data and user-data files or components to exist (see DataSourceNoCloud as an example).
 
The meta-data should really be a yaml dict that minimally contains an unique `instance-id` which cloud-init caches and acts as an indicator to cloud-init that it has applied the desired configuration from the Aeza.net, if that instance-id changes, cloud-init then knows desired configuration has changed for this instance, and it needs to read and apply new configuration for this node. This includes re-creating SSH host keys, setting up default users or whatever other config operations are declared from meta-data/user-data files. 

Other keys in meta-data that are optional: 
- `local-hostname`: the cloud platform's desired hostname for this node
- `cloud-name`: if desired (it will fallback to the `DataSourceAeza.dsname.lower()`)
- `public-keys`: List of public keys to import for the default user
- `region`: string representing the physical region
-  `availability-zone`:  string representing separate AZ if applicable 


Again best practice is to provide minimally `instance-id` that give the platform a knob to trigger a re-run of cloud-init across system boot if necessary due to mandated config/provisioning changes. The rest are optional.

Reply:
> The meta-data should really be a yaml dict that minimally contains an unique instance-id which cloud-init caches

I think we might be being a tad too prescriptive here. At the end of the day, the metadata is useful so that the cloud can get the desired functionality from cloud-init. The form it takes doesn't really matter as long as the cloud-init datasource can scrape it and set the internal structures correctly. Cloud-init can even work just fine without any metadata, but I agree that in order to get cloud-init to do the right thing on snapshots and new instances, a cloud should really be giving us an instance-id...but at the end of the day I wouldn't block on it.